app/src/main/java/it/ministerodellasalute/immuni/logic/exposure/ExposureAnalyticsManager.kt

```diff
@@ -0,0 +1,325 @@
+                isForNextMonth = true
+            )
+        }
+
+        fun scheduleNextDummyInfoReport(serverDate: Date) {
+            val meanWaitingTime = settings.dummyAnalyticsWaitingTime
+            val waitingTime = random.exponential(meanWaitingTime.toLong()).toInt()
+            storeRepository.dummyInfoReportingDate = serverDate.byAdding(seconds = waitingTime)
+        }
+
+        fun couldSendInfo(serverDate: Date): Boolean {
+            return !isWithin24HoursSinceInstall(serverDate)
+        }
+
+        fun hasYetToSendInfoWithExposureThisMonth(serverDate: Date): Boolean {
+            val month = monthFromDate(serverDate)
+            return storeRepository.infoWithExposureLastReportingMonth?.let {
+                it != month
+            } ?: true
+        }
+
+        fun updateInfoWithExposureLastReportingMonth(serverDate: Date) {
+            storeRepository.infoWithExposureLastReportingMonth = monthFromDate(serverDate)
+        }
+
+        fun couldSendInfoWithoutExposureNow(serverDate: Date): Boolean {
+            return isDateWithin24HoursSinceDate(
+                serverDate,
+                storeRepository.infoWithoutExposureReportingDate!!
+            )
+        }
+
+        fun canSendInfoWithExposure(): Boolean {
+            val threshold = settings.operationalInfoWithExposureSamplingRate
+            return random.nextDouble() < threshold
+        }
+
+        fun canSendInfoWithoutExposure(): Boolean {
+            val threshold = settings.operationalInfoWithoutExposureSamplingRate
+            return random.nextDouble() < threshold
+        }
+
+        fun couldSendDummyInfoNow(serverDate: Date): Boolean {
+            return isDateWithin24HoursSinceDate(
+                serverDate,
+                storeRepository.dummyInfoReportingDate!!
+            )
+        }
+
+        private fun isDateWithin24HoursSinceDate(date: Date, periodStartDate: Date): Boolean {
+            return date in periodStartDate..periodStartDate.byAdding(hours = 24)
+        }
+
+        private fun isDatePast24HoursSinceDate(date: Date, periodStartDate: Date): Boolean {
+            return date >= periodStartDate.byAdding(hours = 24)
+        }
+
+        private fun randomDateInMonth(serverDate: Date, isForNextMonth: Boolean): Date {
+            val lastDayOfCurrentMonth = Calendar.getInstance().apply {
+                time = serverDate
+                add(Calendar.MONTH, if (isForNextMonth) 2 else 1)
+                set(Calendar.DAY_OF_MONTH, 1)
+                add(Calendar.DAY_OF_MONTH, -1)
+            }.get(Calendar.DAY_OF_MONTH)
+
+            val randomDayOfCurrentMonth = random.nextInt(lastDayOfCurrentMonth) + 1
+
+            return Calendar.getInstance(TimeZone.getTimeZone("UTC")).apply {
+                time = serverDate
+                if (isForNextMonth) {
+                    add(Calendar.MONTH, 1)
+                }
+                set(Calendar.DAY_OF_MONTH, randomDayOfCurrentMonth)
```

Comment:
if `isForNextMonth` is false, the randomDateOfCurrentMonth could be in the past. Is this desired?

Reply:
Yes, it's by design. In case a time in the past is picked, a call to this method (with isForNextMonth = true) is performed a couple of lines below